Proposal: Introduce Oldtimer

DirectXMan12 opened the pull request:
This is a proposal for Oldtimer, the Heapster historical metrics
access component.  Oldtimer was original proposed in the vision
statement, but was not specified in any particular detail previously.

DirectXMan12 committed:
Proposal: Introduce Oldtimer

This is a proposal for Oldtimer, the Heapster historical metrics
access component.  Oldtimer was original proposed in the vision
statement, but was not specified in any particular detail previously.

ncdc commented on docs/proposals/old-timer.md:
```diff
@@ -0,0 +1,144 @@
+1. Using aggregated historical data to make size-related decisions
+   (for example, idling requires looking for traffic over a long time period)
+
+2. Providing a common interface to for users to view historical metrics
```
remove the "to" after interface

DirectXMan12 replied:
:heavy_check_mark:

ncdc commented on docs/proposals/old-timer.md:
```diff
@@ -0,0 +1,144 @@
+
+2. Providing a common interface to for users to view historical metrics
+
+Before the Heapster refactoring (see the "Heapster Long Term Vision" proposal),
```
Link to proposal?

DirectXMan12 replied:
:check:

ncdc commented on docs/proposals/old-timer.md:
```diff
@@ -0,0 +1,144 @@
+Oldtimer will present an api somewhat similar to the normal Heapster model.
+The urls will take the forms:
+
+`/api/v1/old-timer/{prefix}/metrics/`: Returns a list of all available metrics.
```
I can't say I particularly like "old-timer". `/api/v1/historical` or something along those lines sounds better to me.

DirectXMan12 replied:
Fair.  I like "historical" as well. :heavy_check_mark:

ncdc commented on docs/proposals/old-timer.md:
```diff
@@ -0,0 +1,144 @@
+- Average (arithmetic mean): `/{metric-name}/average`
+- Maximum: `/{metric-name}/max`
+- Minimum: `/{metric-name}/min`
+- Percentile: `/{metric-name}/{number}-perc`
```
Is `perc` currently in use? If not, maybe `pct` or not abbreviating at all would be better?

DirectXMan12 replied:
I was trying to keep it "readable".  IMO, "perc" is more readable than "pct".  I'd prefer either "perc" or not abbreviating it.

ncdc commented on docs/proposals/old-timer.md:
```diff
@@ -0,0 +1,144 @@
+
+- Do the choice of percentiles need to be limited?  InfluxDB and Hawkular
+appear to support arbitrary percentile values in queries, while GCM v3 appears
+to support 99, 95, 50, 5, and OpenTSDB appears to support 50, 75, 90, 95, 99,
```
Suggest ordering both in ascending or descending instead of mixing them

DirectXMan12 replied:
:heavy_check_mark:

piosz commented:
cc @fgrzadkowski

fabxc commented:
As hinted in #1129 I have my concerns about service-level metrics not being clearly separated. Or better, Heapster not holding up to its intent to _not_ be a general purpose monitoring system.

This was previously touched in #665 and I don't feel it ever actually happened. I'm worried that adding a "generic" read-path at this point will lead to Heapster eventually growing a full meta-QL for all its sinks. If it's trying to be the system through which writes happen, it just makes _sense_ to read through it as well.

Aside from the scope-issue, this has many semantical issues as different sinks have different evaluation behaviors (e.g. extrapolation), which can have drastic effects.
Also all these sinks have different query languages. I see a lot of time being spent working on mapping features onto each other and being a general contention point. You touched that point already at the end of your document.

I would be very interested in the use case for this and the benefit within the next 6 months.
As I've heard there are companies with very elaborate scheduling and auto-scaling that hardly had a need for long-term data.
My instinct tells me to defer such features until its absolutely clear what's required.

DirectXMan12 commented:
> I would be very interested in the use case for this and the benefit within the next 6 months.
> As I've heard there are companies with very elaborate scheduling and auto-scaling that hardly had a need for long-term data.
> My instinct tells me to defer such features until its absolutely clear what's required.

Here's an example use case (in fact, this is an actual thing that we'd like to be able to do):

You would like to write an auto-idler component for Kubernetes.  This component would run at some sort of long-term regular interval (say, for instance, every night).  Then, it would, for some set of scalables (e.g. RCs, deployments, etc), check to see if there had been any activity on a particular metric (e.g. network activity or hits on an HTTP server) over the past 24h by using oldtimer.  Then, if they had no traffic, you could idle them, and set up an auto-unidler.

Without oldtimer, you *could* build this component to talk directly to one particular metrics sink, but that would mean that you could not upstream the component, since it would not work with just any Kubernetes deployment.  Alternatively, we could keep longer-term aggregated metrics in Heapster, but the Heapster vision doc *specifically* talks about moving such functionality to a different component (which I think is valuable, for reasons that I talked about a bit in the proposal).

> I'm worried that adding a "generic" read-path at this point will lead to Heapster eventually growing a full meta-QL for all its sinks. If it's trying to be the system through which writes happen, it just makes sense to read through it as well.

I really *do not* want a full query language for Heapster.  This proposal is about a minimal set of aggregations that would be useful for writing Kubernetes components that deal with historical data (e.g. the idling controller talked about above).

> Also all these sinks have different query languages. I see a lot of time being spent working on mapping features onto each other and being a general contention point. You touched that point already at the end of your document.

So, hopefully this came across in the proposal, but if not I can make it clearer there:  This should basically represent the lowest-common-denominator amongst the sinks.  There should **not** be any "this sink does not support feature X, so fake it in the Oldtimer code".  I did spend some time looking at what the different sinks are capable of, but let me know if I missed anything major.

DirectXMan12 commented:
@fabxc you can also look at the original discussion around the Heapster long-term vision document (#769) to see some other proposed use cases for Oldtimer.

mwielgus commented on docs/proposals/old-timer.md:
````diff
@@ -0,0 +1,146 @@
+
+```go
+type MetricAggregationResult struct {
+    Average *MetricResult
````
Is MetricResult the best structure to pass this data?

mwielgus replied:
Would you consider having multiple "MetricAggregationResult" and then a single data point in average, max, min, median, count etc?

DirectXMan12 replied:
I'm definitely open to changes here.  Were you thinking something like:

```go
type MetricAggregationResultList struct {
    Items []MetricAggregationResult
}

type MetricAggregationResult struct {
    Average *MetricPoint
    Maximum *MetricPoint
    ...
}
```

That seems reasonable to me.

DirectXMan12 replied:
Whoops, that last comment needs bucket support:

```go
type MetricAggregationResultList struct {
    Items []MetricAggregationResult
}

type MetricAggregationResult struct {
    Buckets []MetricAggregationBucket
    // this below isn't strictly necessary, but puts the result into context nicely
    BucketSize time.Duration
}

type MetricAggregationBucket struct {
    Average *MetricPoint
    Maximum *MetricPoint
    ...
}
```

DirectXMan12 replied:
Alternatively, the timestamp could be moved into the `MetricAggregationBucket` type,
and have each field just be a numeric value.

mwielgus replied:
Yeah, something like that. Although I would introduce start/end time to MetricAggregationBucket. The question is also whether you want to name all the aggregations upfront or have a map[string]MetricResult (I don't have a super strong opinon on that).

DirectXMan12 replied:
> Although I would introduce start/end time to MetricAggregationBucket

AFAICT, Influx doesn't actually provide an end time -- it just has a timestamp (otherwise I would have).

mwielgus commented on docs/proposals/old-timer.md:
```diff
@@ -0,0 +1,146 @@
+depend mainly on how quickly the sinks can respond to queries.
+
+## Open Questions
+
```
Which sinks are you planning to support?

DirectXMan12 replied:
So, I'd definitely support Hawkular, and I'm also willing to write the initial support for other sinks that people think are important.  I've tried to make sure all the features talked about in the proposal work across InfluxDB, Hawkular, GCM, and OpenTSDB (the main metrics sinks, AFAICT).

DirectXMan12 replied:
The goal is to have a set of features that all the current metrics sinks support, and which seem likely for any future metrics sinks to support.

piosz commented:
@bryk @kubernetes/dashboard-maintainers this is something you are interested in. PTAL

bryk commented:
@piosz @DirectXMan12 I'll review this on behalf of Dashboard UI team.

cheld commented:
CC @taimir

bryk commented on docs/proposals/old-timer.md:
```diff
@@ -0,0 +1,146 @@
+Prior to the Heapster refactor, the Heapster model presented aggregations of
+metrics over certain time periods (the last hour and day).  Post-refactor, the
+concern of presenting an interface for historical metrics was to be split into
+a separate Heapster component: Oldtimer.
```
Will this be built into Heapster? Or maybe it'll be running in a separate Pod/Replication Controller/Service?

DirectXMan12 replied:
So, you'd probably want to run it separately from Heapster so that you could scale it, but we might want to also provide an easy option to run it as part of Heapster as well (some sort of "allinone" installation for quickly getting up and running).

mwielgus replied:
I would suggest to include it in the default deployment. I think that a for 99% deployments a single instance will be enough. It should rather not consume lots of memory. 
I think that anyone who needs to perform large-scale data analysis will do it in some other way (not via this "poor" api).

DirectXMan12 replied:
Sounds good to me.

bryk replied:
LGTM

bryk commented on docs/proposals/old-timer.md:
```diff
@@ -0,0 +1,146 @@
+
+### API
+
+Oldtimer will present an api somewhat similar to the normal Heapster model.
```
Can you highlight the differences? Is this the same API, but with different prefix and bucketing options? Or something else?

DirectXMan12 replied:
Sure, I'll add in a brief overview of the differences.

bryk commented on docs/proposals/old-timer.md:
```diff
@@ -0,0 +1,146 @@
+`/namespaces/{namespace}/pods/{pod-name}/containers/{container-name}`
+(container-level).
+
+In addition, when `{prefix}` is not empty, there will be a url of the form:
```
Out of curiosity, what is this for?

DirectXMan12 replied:
Listing the available items is to be roughly in line with the Heapster model API.  In general, I think both APIs should follow the same conventions.

bryk replied:
All right, sounds good.

bryk commented on docs/proposals/old-timer.md:
```diff
@@ -0,0 +1,146 @@
+
+### Functionality
+
+When Oldtimer receives a request at one of the given URLs, it will compose a
```
Does Oldtimer always talk to a persistent storage? Or, when it knows that everything is in memory already, it returns data from memory? 

Asking to asses performance for lists of pods.

DirectXMan12 replied:
I was envisioning Oldtimer always querying the storage.  Otherwise, how will we know if "everything" is in memory?  Things could have changed since our last fetch.  The way things are tagged in the sinks shouldn't make it *too* bad to list all pods for a namespace, I would image, although I'd welcome additional data/feedback there.

bryk replied:
> Otherwise, how will we know if "everything" is in memory?

@mwielgus does heapster have this info?

DirectXMan12 replied:
Yeah, if it's running in the same process with Heapster, it could just use the info from Heapster.

bryk commented on docs/proposals/old-timer.md:
```diff
@@ -0,0 +1,146 @@
+of (Timestamp, Value) pairs for the requested {prefix}-level metric, over the
+given time range.
+
+`/api/v1/historical/{prefix}/metrics/{metric-name}/{aggregation-name}?start=X&end=Y&bucket=B`:
```
Can you formalize X, Y and B?

Whats the URL to get, say, hourly 95-percentiles for last 30 days?

DirectXMan12 replied:
They're all specified slightly below (after the definition of `{prefix}`), but if that's not clear, I can more the definitions around a bit.  An example probably wouldn't hurt either ;-)

Hourly 95-percentiles from the last 30 days: `GET /api/v1/historical/namespaces/default/pods/somepod/metrics/cpu-usage/95-perc?start=2016-03-20T10:57:37-04:00&bucket=1h`

bryk replied:
All right, thanks. Sounds good.

bryk commented:
Looks nice. A few questions.

DirectXMan12 commented:
Ok, I've tried to address all of the comments.  I've added in an example aggregation query, and made a few more things more explicit, and rearranged the proposal a bit.

bryk commented:
LGTM from Dashboard UI perspective

mwielgus commented on docs/proposals/old-timer.md:
```diff
@@ -0,0 +1,212 @@
+concern of presenting an interface for historical metrics was to be split into
+a separate Heapster component: Oldtimer.
+
+Oldtimer will run as part of the main Heapster executable, and will present
```
Oldtimer should rather run as a separate executable/containert in the main Heapster pod.

DirectXMan12 replied:
ah, ok, I think I misunderstood you above.

DirectXMan12 replied:
That would effectively mean that it would have to be on a different port.  Why not just allow it to run in the same executable, if you're going to be running it as part of the same pod anyway, so that `https://heapster.kube-system/api/v1/model` is for the model, while `https://heapster.kube-system/api/v1/historical` is for oldtimer?

DirectXMan12 commented:
Had to tweak the paths a bit so that go-restful could work with them (it turns out go-restful doesn't like wildcards in the middle of a path, and it's probably not the best idea as far as ambiguity is concerned, anyway)

piosz commented on docs/proposals/old-timer.md:
```diff
@@ -0,0 +1,212 @@
+Several different aggregations will be supported.  Aggregations should be
+performed in the metrics sink.  If more aggregations later become supported
+across all metrics sinks, the list can be expanded (and the API version
+should probably be bumped, since the supported aggregations should be part of
```
Please note that we actually don't have versioned API in Heapster. This a kind of promise that we will try to support it in some backward-compatible way but there is no guarantee. If you really want to have a versioned API it should be a part of Resource Metrics API https://github.com/kubernetes/kubernetes/pull/24253

DirectXMan12 replied:
I'll tweak the wording here.  I figured since there was an API version in the path, the intention was that you'd have a versioned API at some point.

piosz replied:
There will be another versioned API of Kubernetes standards (proposal linked in previous comment). Especially if you want model/historical api to be fully versioned they should go through the path: alpha -> beta -> stable. We don't want it here to allow higher velocity at cost of possible instability.

piosz commented:
Let's imagine the situation where we created a pod with some name, it was deleted and then let's say a week ago we created another pod with the same name which is a totally different pod. What will be the semantic of historical metrics in this case? Do you plan to verify also pod-uid somehow?

DirectXMan12 commented:
> What will be the semantic of historical metrics in this case? Do you plan to verify also pod-uid somehow?

Good point.  We probably want `/pod-ids/{pod_id}` and `/pod-ids/{pod_id}/containers/{container}` API endpoints.  Keeping the namespace-pod API endpoint as well might be nice, though, since it mirrors the model API.

piosz commented:
But operating on pod uids is a bad user experience.

I'm ok with saying (possibly in the first version): if you have two pods with the same name historically the data will be mixed. Better approach is to return data only from the newest pod with the given name. For example I created on Monday pod with name `my-name` and then killed it on Tuesday. I created also a pod with the same name on Friday and it is still running. When querying for historical data for the whole week I'd like to get only metrics for the second pod, so there won't be any data for Mon-Thu. WDYT?

DirectXMan12 commented:
> Better approach is to return data only from the newest pod with the given name

This gets a bit complicated, since effectively we have to then know the appropriate pod uid for a pod name *before* making a query, so we have to make two queries against the database.  That might not be so bad, but it does adds another round trip for every query.

If we're ok with the extra round trip, the optimal setup would probably be that you can refer to pod UID to get a specific pod, or refer to pod name to get the newest pod with that name.

DirectXMan12 commented:
I've added in the pod-id-based endpoints, and added a note about the pod-name-based endpoints.  I think there are some cases for certain backends in which we can get away with only one trip to the backend, but there are several cases where two will be required.

For instance, to get raw metrics from Hawkular, you *need* to know the pod id, AFAICT (@mwringe et all might be able to correct me on this).  For aggregations, I'm not sure how Hawkular deals with a query which could return multiple time series, but I suspect we'll need it there as well.

DirectXMan12 commented:
I've also clarified a couple points for posterity's sake:

- Oldtimer is intended to work only with data written in the Heapster storage schema (i.e. data originally written by Heapster)
- Oldtimer is not mean to be a general-purpose query tool that maps arbitrary queries to the sink query language -- it's just supposed to support a basic set of aggregations on single time series.

cc @fabxc 

I think at this point the proposal is pretty much ready for merge.  @mwielgus anything else you want changed?

mwielgus commented:
LGTM

mwringe commented:
> For instance, to get raw metrics from Hawkular, you need to know the pod id, AFAICT (@mwringe et all might be able to correct me on this). For aggregations, I'm not sure how Hawkular deals with a query which could return multiple time series, but I suspect we'll need it there as well.

Hmm, you shouldn't need to know the pod id to get the raw data, you should be able to do a query based on any of the labels. But there may have been an issue preventing this in the past.

Can you get in touch with Stefan Negrea (he is stefan_n in #hawkular on freenode) to answer your questions?
